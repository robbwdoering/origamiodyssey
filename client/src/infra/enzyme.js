/**
 * FILENAME: enzyme.js
 *
 * DESCRIPTION: Setup for AirBnB's Enzyme package. 
 */

import Enzyme, { configure, shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
export { shallow, mount, render };
export default Enzyme;
