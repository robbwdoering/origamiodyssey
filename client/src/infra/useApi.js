/**
 * FILENAME: useApi.js
 *
 * DESCRIPTION: Helper hook that manages a connection to the authenticated REST API that backs the app.
 * src: https://github.com/auth0/auth0-react/blob/master/EXAMPLES.md#4-create-a-useapi-hook-for-accessing-protected-apis-with-an-access-token
 * @param path JUST THE PATH, not the full URL, of the requested resource
 * @param options object defining audience, scope, and any custom inputs
 */
import { useState } from 'react';
import { useAuth0 } from '@auth0/auth0-react';

export const useApi = (url, method, options = {}, handleFetch, data) => {
    const { getAccessTokenSilently, user, isAuthenticated } = useAuth0();
    const [state, setState] = useState({
        error: null,
        loading: false,
        data: null
    });

    const sleep = ms => {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    const perform = async (urlSuffix, overrideValues, overrideMethod) => {
        let count = 20;
        while ((!isAuthenticated) && count > 0) {
            count--;
            await sleep(1000);
        }

        try {
            const accessToken = await getAccessTokenSilently(options);

            let msg = {
                method: overrideMethod || method,
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                    // Authorization: `Bearer ${accessToken}`,
                    strategy: 'auth0',
                    access_token: accessToken
                }
            };

            // Append data if it was passed
            if (overrideValues) {
                msg.body = JSON.stringify(overrideValues);
            } else if (data && method !== "GET") {
                msg.body = JSON.stringify(data);
            }
            setState(Object.assign({}, state, { loading: true }))

            // Perform the fetch asynchrounously
            fetch(process.env.REACT_APP_API_IP + url + (urlSuffix || ""), msg)
            .then(response => {
                if (!response.ok) {
                    console.error('[useApi] Received a bad status from the server: ', response);
                    throw new Error(`status ${response.status}`);
                }
                return response.json();
            })
            .then((json) => {
                // This line handles all responses - created at app root to hook into redux
                if (handleFetch) {
                    handleFetch(json);
                }

                setState(Object.assign({}, state, {data: json}))

                // Turn off loading indicator
                setState(Object.assign({}, state, { loading: false }));
            })
            .catch(e => {
                console.log('call failed', e);
            });
        } catch (error) {
            console.log('useApi FAILURE from exception: ', error);
            setState({
                ...state,
                error,
                loading: false
            });
        }
    };

    return {
    ...state,
    refresh: perform
    };
};
