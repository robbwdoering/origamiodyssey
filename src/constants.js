/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

export const Pages = {
	Splash: "Splash",
	Lesson: "Lesson"
};

export const initAppReducerState = {
	page: Pages.Splash
};
