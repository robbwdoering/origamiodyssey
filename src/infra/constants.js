/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

import React, { useState, useRef, useMemo, useEffect } from 'react';
import GitHubIcon from "@material-ui/icons/GitHub";
import SchoolIcon from "@material-ui/icons/School";
import ClassIcon from "@material-ui/icons/Class";
import DashboardIcon from "@material-ui/icons/Dashboard";
import GroupAddIcon from "@material-ui/icons/GroupAdd";

// The FOLD files 
import BirdBase from "./../folds/BirdBase.json";
import BoatBase from "./../folds/BoatBase.json";
import FrogBase from "./../folds/FrogBase.json";
import SailBoat from "./../folds/SailBoat.json";

export const Pages = {
	Splash: "Splash",
	ModelSelect: "ModelSelect",
	Fold: "Fold",
	User: "User",

	// Used but undefined
	Lesson: "Lesson",
	History: "History",
	LearnDashboard: "LearnDashboard",
	TeachDashboard: "TeachDashboard"
};

export const Tags = {
	base: {
		text: "Base",
		category: "family"
	},
	bird: {
		text: "Bird",
		category: "counterpart",
	},
	boat: {
		text: "Boat",
		category: "counterpart"
	},
	flower: {
		text: "Flower",
		category: "counterpart"
	}
};

export const Folds = {
	BirdBase: {
		name: "Bird Base",
		description: "The first base many learn, this is perfect for winged creations such as birds and dragons.",
		tags: ["base"],
		author: "Traditional",
		json: BirdBase
	},
	BoatBase: {
		name: "Boat Base",
		description: "A more specific base for blocky models, such as boats and frames.",
		tags: ["base"],
		author: "Traditional",
		json: BoatBase
	},
	FrogBase: {
		name: "Frog Base",
		description: "A complex base used for models with length-wise asymmetry, like flowers and animals.",
		tags: ["base"],
		author: "Traditional",
		json: FrogBase 
	},
	SailBoat: {
		name: "Sail Boat",
		description: "An assymetrical sailboat with a built in kickstand - perfect for display.",
		tags: ["boat"],
		author: "Traditional",
		json: SailBoat 
	}
};

export const initAppReducerState = {
	layoutState: {
		hash: 0,
		page: Pages.ModelSelect,
		curFold: null,
		foldHash: 0,
		useImages: false,
		showEditor: true,
		expandHierarchy: false 
	},
	foldState: {
		hash: 0,
		selectedLevel: 0,
		stepIdx: -1, // Depends on selected level
		overrideWithEditor: false,
		usingDefaults: true
	},
	editorState: {
		hash: 0,
		edgeHighlights: [],
		vertexHighlights: [],
		faceHighlights: [],
		showEdges: true,
		showVertices: true,
		showFaces: true,
		showTriangulations: false,
		showLables: false
	},
	showNavDrawer: false

};

export const Actions = {
	SET_LAYOUT_STATE: "SET_LAYOUT_STATE",
	SET_SHOW_NAV_DRAWER: "SET_SHOW_NAV_DRAWER",
	SET_PAGE_CONFIG: "SET_PAGE_CONFIG",
	SET_FOLD_STATE: "SET_FOLD_STATE",
	SET_EDITOR_STATE: "SET_EDITOR_STATE",
};

export const initNavTree = [
	{	
		text: "What is Origami?",
		key: "context_root",
		children: [
			{
				text: "History",
				key: "history"				
			},
			{
				text: "Origami Today",
				key: "current_origami"
			},
			{
				text: "Practical Origami",
				key: "practical"
			}
		]
	},
	{	
		text: "Learn Origami",
		key: "learn_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Dashboard",
				key: "learn_dashboard"				
			},
			{
				icon: <GroupAddIcon />,
				text: "Join Class",
				key: "join_class"
			}
		]
	},
	{	
		text: "Teach Origami",
		key: "teach_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Lesson Plans",
				key: "lesson_plan_dashboard"				
			},
			{
				icon: <SchoolIcon />,
				text: "Read Research Paper",
				key: "google_scholar"
			},
			{
				icon: <ClassIcon />,
				text: "Origami in the Classroom",
				key: "teaching_tips"
			}
		]
	},
	{	
		icon: <GitHubIcon />,
		text: "Source Code",
		key: "github"
	},
];

export const initFoldState = {
	//place to store buffer geo vertex data
	positions: [],
	//place to store buffer geo vertex colors
	colors: [],
	indices: [],
	nodes: [], 
	faces: [], 
	edges: [], 
	creases: [], 
	vertices: [], //indexed vertices array
	fold: null,
	creaseParams: null
}

/*
export const Lines = {
    U: hingeLines,
    M: mountainLines,
    V: valleyLines,
    C: cutLines,
    F: facetLines,
    B: borderLines
};
*/
