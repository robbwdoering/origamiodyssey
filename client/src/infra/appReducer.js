/**
 * FILENAME: appReducer.js
 *
 * DESCRIPTION: Handles state update for all layout actions.
 */

import supermemo2 from 'supermemo2';
import { initAppReducerState, Actions } from './constants';

const finalInitState = JSON.parse(JSON.stringify(initAppReducerState));

export const appReducer = (state = finalInitState, action) => {
	let newState = Object.assign({}, state);

	switch (action.type) {
		case Actions.SET_SHOW_NAV_DRAWER:
			// If passed a val set to that, otherwise toggle
			newState.showNavDrawer = action.payload !== undefined ? action.payload : !newState.showNavDrawer;
			break;
		case Actions.SET_LAYOUT_STATE:
			console.log('[SET_LAYOUT_STATE]', action.payload);
			Object.assign(newState.layoutState, action.payload || initAppReducerState.layoutState);
			newState.layoutState.hash++;
			break;
		case Actions.SET_FOLD_STATE:
			console.log('[SET_FOLD_STATE]', action.payload);
			Object.assign(newState.foldState, action.payload || initAppReducerState.foldState);
			newState.foldState.hash++;
			break;
		case Actions.SET_EDITOR_STATE:
			console.log('[SET_EDITOR_STATE]', action.payload);
			Object.assign(newState.editorState, action.payload || initAppReducerState.editorState);
			newState.editorState.hash++;
			break;
		case Actions.SET_USER_STATE:
			console.log('[SET_USER_STATE]', action.payload);
			Object.assign(newState.userState, action.payload || initAppReducerState.userState);
			newState.userState.hash++;
			break;
		case Actions.ADD_HISTORY_ENTRY:
			if (!action.payload) {
				return newState;
			}
			newState.userState.foldHistory.push(action.payload)

			// If this model type is part of the memo algorithm, update that entry 
			const idx = newState.userState.modelList.findIndex(model => model.foldKey === action.payload.foldKey);
			if (idx !== -1) {
				console.log("Modifying entry for ", action.payload.foldKey);

				// Update the spaced learning params given this new information
				const quality = action.payload.quality;
				const lastSchedule = newState.userState.modelList[idx].schedule
				const lastFactor = newState.userState.modelList[idx].factor
				newState.userState.modelList[idx] = {
					foldKey: action.payload.foldKey,
					...supermemo2(quality, lastSchedule, lastFactor)
				}
			}
			// Else don't rememeber anything other than the foldHistory row

			newState.userState.hash++;
			console.log('[ADD_HISTORY_ENTRY]', action.payload, newState.userState);
			break;
	}

	return newState;
};
