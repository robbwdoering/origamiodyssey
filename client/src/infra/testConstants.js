/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {
    foldState: Object.assign({}, initAppReducerState.foldState, {
        maxSteps: 10
    })
});
