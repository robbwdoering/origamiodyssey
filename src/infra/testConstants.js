/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {});
