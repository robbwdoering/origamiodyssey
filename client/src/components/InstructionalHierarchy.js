/**
 * FILENAME: InstructionalHierarchy.js
 *
 * DESCRIPTION: These are piecemeal controls for the fold state.
 */

// React + Redux
import React, { useState, useRef, useMemo, useEffect, createRef } from 'react';
import { connect } from 'react-redux';

import { useUpdate, useSpring, useSprings, animated, config } from 'react-spring';
import SquareLoader from 'react-spinners/SquareLoader';

import {
	SwipeableDrawer,
	Tooltip,
	Typography,
	Fab,
	ButtonGroup,
	List,
	Divider,
	ListItem,
	Card
} from '@material-ui/core';
import SkipPrevious from '@material-ui/icons/SkipPrevious';
import SkipNext from '@material-ui/icons/SkipNext';
import ExpandMore from '@material-ui/icons/ExpandMore';
import ExpandLess from '@material-ui/icons/ExpandLess';

import useStyles from './../style/theme';
import { Folds } from './../infra/constants';
import { setFoldState, setLayoutState, setUserState } from './../infra/actions';
import TimerContainer from './Timer';
import { collectStepsForLevel, calcMaxLevel, printPath, findInUseFamilyNode, getHierNode } from './../infra/utils';

const HIER_PX_SIZE = 20;

export const InstructionalHierarchy = props => {
	const {
		initFold,
		foldLastUpdated,
		foldState,
		setFoldState,
		layoutState,
		setLayoutState,
		userState,
		setUserState
	} = props;

	// ----------
	// STATE INIT
	// ----------

	const [contStyle, setContStyle] = useState({});
	const classes = useStyles();
	const cardRef = useRef();
	const activeNodeRef = useRef();
	const renderRows = useRef([]);
	const [curHash, setHash] = useState(0);
	const [looperDirection, setLooperDirection] = useState(-1);
	const [looperWorkerId, setLooperWorkerId] = useState(-1);
	const [looperHash, setLooperHash] = useState(0);
	const [finishedInit, setFinishedInit] = useState(false);

	const maxLevel = useMemo(() => calcMaxLevel(initFold && initFold.instructions), [foldLastUpdated]);

	const trackTop = window.innerHeight - 264 - HIER_PX_SIZE * foldState.stepIdx + 'px';

	// ----------------
	// MEMBER FUNCTIONS
	// ----------------

	/**
	 * Changes the current instructional sequential step, prompting animation through redux changes.
	 * @param delta The number of step indices to change by. Positive to go towards end of instructions, negative to go back
	 * @param isFromLooper true if this step change was done automatically
	 */
	const changeStep = (delta, isFromLooper) => {
		let newStepIndex = Math.min(Math.max(foldState.stepIdx + delta, -1), foldState.maxSteps);
		let newFoldState = {
			stepIdx: newStepIndex
		};

		if (!isFromLooper) {
			newFoldState.repeatRoot = -1;
			newFoldState.repeatRange = null;
			if (looperWorkerId !== -1) {
				clearInterval(looperWorkerId);
				setLooperWorkerId(-1);
			}
		}

		setFoldState(newFoldState);
	};

	/**
	 * Calculate how many "default", or 2D, steps come before this one, and get their lengths.
	 * This is useful since we know our index within the stepArray, but without information from this function,
	 * we would have no way of deriving the final height of the whole component.
	 * @param stepIdx count all the steps before this index
	 */
	const calcNumDefaultsBefore = stepIdx => {
		let total = 0;

		stepArray.some((elem, i) => {
			if (i > stepIdx) {
				return true;
			}

			// Only default items are 2D arrays
			if (Array.isArray(elem[1][0])) {
				// Add the number of leaf nodes here, minus one for the path at idx 0, and one for the spot taken by the root 2D array
				total += elem.length - 2;
			}
		});

		return total;
	};

	/**
	 * Calculate where the card appears based on the current step and the shape of the overall instructions.
	 * @return a CSS-in-JS styling object
	 */
	const calcCardPos = () => {
		// const right = window.innerWidth > 1200 ? ((window.innerWidth / 2) + 0 + 'px') : undefined;
		// const left = right ? undefined : 10;
		const width = (maxLevel - 1) * HIER_PX_SIZE + 14;

		// Left side is outside the center column if there's room, else just sticks to the left side
		const left = window.innerWidth < 1200 + width + 10 ? 0 : -width - 10;

		// Top updates every time the step changes
		const numDefaultsBefore = calcNumDefaultsBefore(foldState.stepIdx);
		const top = window.innerHeight - 296 - HIER_PX_SIZE * (foldState.stepIdx + 1 + numDefaultsBefore);
		let style = {
			// Don't show first column, and account for padding
			width: width + 'px',
			left: left + 'px',
			height: contStyle.height,
			top: top + 'px'
		};

		return style;
	};

	/**
	 * Rerender this component.
	 */
	const triggerRerender = () => {
		setHash(cur => cur + 1);
	};

	/**
	 * Execute logic necessary when a user clicks one of the nodes, which usually will simply set the
	 * current step to whatever node was clicked, but can do other things if, e.g., they held shift to
	 * start a loop.
	 * @param event Event object from MUI
	 * @param path Unique string representing the node that was clicked
	 */
	const handleHierNodeClick = (event, path) => {
		const stepIdx = stepArray.findIndex(step => path === step[0]);
		let newFoldState = {};

		const [startUseIndex, endUseIndex] = findInUseFamilyNode(stepArray, path);

		if (!event.shiftKey) {
			// If no shift, just move to this node
			if (stepIdx !== -1) {
				// If it's in the stepArray, just set the index to that element
				newFoldState.stepIdx = stepIdx - 1;
			} else {
				// Else we need to find whatever in use index corresponds
				if (startUseIndex !== -1 && endUseIndex !== -1) {
					newFoldState.stepIdx = startUseIndex - 1;
				}
			}

			if (foldState.repeatRoot !== -1) {
				newFoldState.repeatRoot = -1;
				newFoldState.repeatRange = null;

				if (looperWorkerId !== -1) {
					clearInterval(looperWorkerId);
					setLooperWorkerId(-1);
				}
			}
		} else {
			// If shift held, then start or modify repeatRange
			if (foldState.repeatRoot !== -1) {
				// Modify existing range, replacing the second clicked index
				if (stepIdx !== -1) {
					newFoldState.repeatRange = [foldState.repeatRoot, stepIdx];
				} else if (startUseIndex <= foldState.repeatRoot && endUseIndex >= foldState.repeatRoot) {
					// Either clicked a descendant or ancestor of the root - basically undefined behavior
					newFoldState.repeatRange = [startUseIndex, endUseIndex];
				} else if (startUseIndex < foldState.repeatRoot) {
					// Clicked something before
					newFoldState.repeatRange = [foldState.repeatRoot, startUseIndex];
				} else {
					// Clicked something after
					newFoldState.repeatRange = [foldState.repeatRoot, endUseIndex];
				}
			} else {
				if (stepIdx !== -1) {
					// If this is the first shift click, we're starting at this location
					newFoldState.stepIdx = stepIdx - 1;

					newFoldState.repeatRoot = stepIdx;
					newFoldState.repeatRange = [stepIdx, stepIdx];
				} else {
					// If this is the first shift click, we're starting at this location
					newFoldState.stepIdx = startUseIndex - 1;

					newFoldState.repeatRoot = startUseIndex;
					newFoldState.repeatRange = [startUseIndex, endUseIndex];
				}
			}

			if (looperWorkerId === -1) {
				setLooperWorkerId(setInterval(looperWorker, 2500));
			}

			// Ensure the first element is always the first in index order
			newFoldState.repeatRange.sort();

			// Go to the start, and progress forward
			newFoldState.stepIdx = newFoldState.repeatRange[0] - 1;
			setLooperDirection(1);
		}

		setFoldState(newFoldState);
	};

	/**
	 * Wrapper for collectStepsForLevel using local vars.
	 * @returns the array of "actual"/"in-use" steps
	 */
	const buildStepArray = () => {
		return collectStepsForLevel(initFold, 0, foldState.usingDefaults);
	};

	/**
	 * Render one node in the hierarchy.
	 * NOTE: This doesn't return the node, but rather saves it to the renderRows object passed in.
	 * @param inst this particular instruction object
	 * @param renderRows output array of renderable objects
	 * @param levelIdx the index of the current hierarchy level
	 * @param path the path of this instructional node
	 * @param belowDefault true if we are "below" a default node, which requires special handling
	 * @return the height of the rendered row
	 */
	const renderNode = (inst, renderRows, levelIdx, path, belowDefault) => {
		// Error case - we're on a level deeper than the tree supports
		if (levelIdx > maxLevel) {
			return [1, stepIdx];
		}
		let height = 0;
		let children = [];
		const isDefaultNode = foldState.usingDefaults && inst.default;
		const isSelectedLevel = isDefaultNode || !inst.children.length;
		const isLeaf = inst.children.length && Array.isArray(inst.children[0]);
		const stepIdx = stepArray.findIndex(step => path === step[0]);

		const style = {};

		if (isLeaf) {
			// Base case: This is a leaf node
			height = 1;

			style.flexGrow = height;

			// Add spacers to all rows below, if necessary
			for (let i = levelIdx + 1; i < maxLevel; i++) {
				renderRows[i].push(
					<div className={`${classes.hier_node_anchor}`} style={style}>
						<div className={`${classes.hier_node} ${classes.hier_node_spacer}`} key={path + 'spacer' + i} />
					</div>
				);
			}
		} else if (inst.children.length) {
			// Recursive case: this node has children, so it's as tall as all of them combined
			[height, children] = inst.children.reduce(
				(acc, child, childIdx) => {
					const childHeight = renderNode(
						child,
						renderRows,
						levelIdx + 1,
						`${path},${childIdx}`,
						belowDefault || isDefaultNode
					);
					acc[0] += childHeight;
					return acc;
				},
				[0, []]
			);

			style.flexGrow = height;
		}

		let type = 'default';

		if (stepIdx === foldState.stepIdx + 1) {
			// This node is the current step (i.e. haven't folded yet)
			type = 'active';
		} else if (isSelectedLevel || (isLeaf && !belowDefault)) {
			// These nodes are part of the current step array
			type = 'inUse';
		}

		const pxHeight = HIER_PX_SIZE * style.flexGrow - 4;

		renderRows[levelIdx].push(
			<Tooltip
				title={inst.desc}
				placement="bottom-start"
				classes={{ popper: classes.hier_node_tooltip }}
				key={path}
			>
				<div
					key={`${path}-div`}
					className={`${classes.hier_node_anchor}`}
					ref={activeNodeRef}
					style={style}
					onClick={event => handleHierNodeClick(event, path)}
				>
					<div
						key={`${path}-div-div`}
						className={`${classes.hier_node} ${classes['hier_node__' + type]}`}
						style={{ height: pxHeight }}
					/>
				</div>
			</Tooltip>
		);

		return height;
	};

	/**
	 * Re-render all the rows.
	 * @param cardStyle the style object that describes the shape of the component
	 */
	const refreshRenderRows = cardStyle => {
		if (!initFold) {
			return;
		}

		let tmpRows = [];
		for (let i = 0; i < maxLevel; i++) {
			tmpRows.push([]);
		}

		renderNode(initFold.instructions, tmpRows, 0, '0');

		renderRows.current = tmpRows;
		const longestRowLen = Math.max(...tmpRows.map(arr => arr.length));
		setContStyle(
			Object.assign({}, contStyle, {
				height: HIER_PX_SIZE * longestRowLen,
				width: cardStyle.width
			})
		);
	};

	/**
	 * Set the current step to the last one, ending the instructions.
	 */
	const jumpToEnd = () => setFoldState({ stepIdx: foldState.maxSteps - 1 });

	/**
	 * Get the description for the current step.
	 * @param stepIdx the index of the current step
	 */
	const getDescForNode = stepIdx => {
		const step = stepArray[foldState.stepIdx + 1];
		const path = step[0].split(',').slice(1);
		let node = getHierNode(initFold.instructions, path);

		return node.desc;
	};

	/**
	 * Changes the looperHash, which runs updateLooper() via a lifecycle call.
	 */
	const looperWorker = () => {
		if (foldState.repeatRoot === -1 || !foldState.repeatRange) {
			console.log('[looperWorker] ERR - null arguments');
			return;
		}

		setLooperHash(curHash => curHash + 1);
	};

	/**
	 * Carry out one iteration of the looping behavior, either going one forward one step or jumping back to the start of the loop.
	 */
	const updateLooper = () => {
		if (foldState.repeatRoot === -1 || !foldState.repeatRange) {
			return;
		}

		const distFromStart = foldState.stepIdx - foldState.repeatRange[0];
		const distFromEnd = foldState.repeatRange[1] - foldState.stepIdx;

		// Switch direction to go back up
		let tmpDirection = looperDirection;

		if (tmpDirection === -1 && distFromStart === -1) {
			tmpDirection = 1;
			setLooperDirection(tmpDirection);
		} else if (tmpDirection === 1 && distFromEnd === 1) {
			tmpDirection = -1;
			setLooperDirection(tmpDirection);
		}

		changeStep(tmpDirection, true);
	};

	/**
	 * Render the looper indication, which is a series of boxex, one for each step already executed for this loop.
	 * @return a renderable object
	 */
	const renderLooperItems = () => {
		if (foldState.repeatRoot === -1 || !foldState.repeatRange) {
			return null;
		}

		let ret = [];
		for (let i = foldState.repeatRange[0]; i <= foldState.repeatRange[1]; i++) {
			ret.push(
				<div
					key={i}
					className={`${classes.hier_looper_item} ${
						i <= foldState.stepIdx + 1 ? classes.hier_looper_item__active : ''
					}`}
				/>
			);
		}

		return ret;
	};

	// ---------
	// LIFECYCLE
	// ---------

	// On mount build the button classes object
	const buttonClasses = useMemo(
		() => ({
			root: classes.fold_controls_button,
			label: classes.fold_controls_button_label
		}),
		[]
	);

	// Build a new "actual"/"in-use" step array whenever the instructions change
	const stepArray = useMemo(buildStepArray, [!initFold || !initFold.instructions, initFold && initFold.frame_title]);

	// Recalculate the shape of the card every render
	const cardStyle = calcCardPos();

	// Perform mount and unmount actions
	useEffect(() => {
		// Rerender whenever the page resizes
		window.addEventListener('resize', triggerRerender);

		// Unmount logic
		return () => {
			// Stop the looper worker if it's running
			if (looperWorkerId !== -1) {
				clearInterval(looperWorkerId);
			}
		};
	}, []);

	// Start actually rendering text after a second
	useEffect(() => setTimeout(() => setFinishedInit(true), 1000), []);

	// Rerender the data every time the step index changes (or the fold itself changes)
	useEffect(() => refreshRenderRows(maxLevel), [foldState.stepIdx, foldLastUpdated]);

	// Take another step in the looping behavior whenever asked to
	useEffect(updateLooper, [looperHash]);

	// Calculate all these every render
	const ctrlCardLeftPx = `${window.innerWidth / 2 + 256}px`;
	const curBodyWidth = Math.min(1200, window.innerWidth);
	const buttonSize = window.innerWidth > 440 ? Math.min(200, curBodyWidth * 0.15) : 95;
	const buttonIconClass =
		buttonSize < 100 ? classes.fold_controls_button_icon : classes.fold_controls_button_icon_large;
	const fabStyle = {
		width: buttonSize + 'px',
		height: buttonSize + 'px'
	};

	// Magic nums:
	// 1200 is max width of the center colun
	// 10 is common spacing
	// 32 is padding on center column
	const instCardMargin =
		window.innerWidth < 1200 + parseInt(cardStyle.width) + 10 ? parseInt(cardStyle.width) - 32 + 10 + 'px' : '0';

	return (
		<div key='hier-center-col' className={classes.centerColumn_flex}>
			{/* The card contains the timeline, which contains most actions here */}
			<div key='hier-card' className={classes.hier_card} style={cardStyle} ref={cardRef}>
				{/* The timeline */}
				<div key='hier-timeline' className={classes.hier_container} style={contStyle}>
					{initFold &&
						renderRows.current.reduce((acc, row, idx) => {
							if (idx !== 0) {
								acc.push(
									<div idx={idx} key={'oo-hcr-' + idx} className={classes.hier_node_container}>
										{row.map((elem, subIdx) => <div key={'oo-oo-hcr-' + subIdx}> {elem} </div>)}
									</div>
								);
							}
							return acc;
						}, [])}
					<div className={`${classes.hier_node} ${classes.hier_node_bookend}`} />
				</div>

				{/* The "current Time" line */}
			</div>

			{/* Text box shows details on the current step */}
			{initFold && initFold.instructions && (
				<div className={classes.hier_desc_container}>
					{foldState.repeatRoot !== -1 && (
						<div className={classes.hier_looper_rail} style={{ marginLeft: instCardMargin }}>
							<div className={classes.hier_looper_container}>{renderLooperItems()}</div>
						</div>
					)}
					<Card className={classes.hier_desc_card} style={{ marginLeft: instCardMargin }}>
						{finishedInit ? (
							<React.Fragment>
								<Typography className={classes.modelCard_title} variant="h5" component="h2">
									{' '}
									Current Step{' '}
								</Typography>
								<Typography>
									{foldState.stepIdx < foldState.maxSteps - 1
										? getDescForNode(foldState.stepIdx)
										: // "text" :
										  'Congratulations - your model is complete!'}
								</Typography>
							</React.Fragment>
						) : (
							<SquareLoader color="#e0e0e0" size={75} css="margin: 16px;" />
						)}
					</Card>
				</div>
			)}

			<div className={classes.fold_controls_button_container} id="oo-main-button-step-ctrls">
				<Fab
					classes={buttonClasses}
					onClick={() => changeStep(-1)}
					disabled={foldState.stepIdx < 0}
					style={fabStyle}
					color="primary"
					size="large"
				>
					<SkipPrevious className={buttonIconClass} />
				</Fab>
				<Fab
					classes={buttonClasses}
					onClick={() => changeStep(1)}
					disabled={foldState.stepIdx >= foldState.maxSteps - 1}
					style={fabStyle}
					color="primary"
					size="large"
				>
					<SkipNext className={buttonIconClass} />
				</Fab>
			</div>

			{userState.showTimerAssess && <TimerContainer />}
		</div>
	);
};

export const mapStateToProps = (state, props) => {
	return {
		layoutState: state.appReducer.layoutState,
		layoutStateHash: state.appReducer.layoutState.hash,
		foldState: state.appReducer.foldState,
		foldStateHash: state.appReducer.foldState.hash,
		userState: state.appReducer.userState,
		userStateHash: state.appReducer.userState.hash
	};
};

export default connect(mapStateToProps, { setFoldState, setLayoutState, setUserState })(InstructionalHierarchy);
