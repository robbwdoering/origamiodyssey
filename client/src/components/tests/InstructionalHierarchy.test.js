/**
 * FILENAME: InstructionalHierarchy.test.js
 */

// React + Enzyme
import React from 'react';
import { mount } from './../../infra/enzyme';

// Local
import { testRedux } from './../../infra/testConstants';
import { Folds } from './../../infra/constants';
import { InstructionalHierarchy } from './../InstructionalHierarchy';

describe('Fold Editor', () => {
	let comp;
	let setFoldState, setLayoutState, setUserState;

	beforeEach(() => {
		setFoldState = jest.fn();
		setLayoutState = jest.fn();
		setUserState = jest.fn();

		comp = mount(
			<InstructionalHierarchy
				windowHeight={500}
				initFold={Object.assign({}, Folds.BirdBase.json)}
				foldLastUpdated={0}
				foldState={Object.assign({}, testRedux.foldState)}
				foldStateHash={0}
				layoutState={Object.assign({}, testRedux.layoutState)}
				layoutStateHash={0}
				userState={Object.assign({}, testRedux.userState)}
				userStateHash={0}
				setFoldState={setFoldState}
				setLayoutState={setLayoutState}
				setUserState={setUserState}
			/>
		);
	});

	afterEach(() => {
		comp.unmount();
	});

	it('renders without crashing', () => {
		expect(comp).toMatchSnapshot();
	});

	it('changes step when the buttons are used', () => {
		// Advance one
		comp.find('#oo-main-button-step-ctrls').find('button').last().simulate('click');

		expect(setFoldState).toHaveBeenCalled();
		expect(setFoldState.mock.calls[0][0].stepIdx).toBe(0);

		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				stepIdx: 5
			})
		});
		comp.update();
		comp.find('#oo-main-button-step-ctrls').find('button').first().simulate('click');

		expect(setFoldState.mock.calls.length).toBe(2);
		expect(setFoldState.mock.calls[1][0].stepIdx).toBe(4);
	});

	it('changes step when a node is clicked on', () => {
		comp.find('ForwardRef(Tooltip)').at(5).simulate('click');

		expect(setFoldState).toHaveBeenCalled();
		expect(setFoldState.mock.calls[0][0].stepIdx).toBe(3);
	});

	it('enables looping behavior', () => {
		// Turn on looper
		comp.find('ForwardRef(Tooltip)').at(5).simulate('click', { shiftKey: true });

		expect(setFoldState).toHaveBeenCalled();
		expect(setFoldState.mock.calls[0][0]).toEqual({
			repeatRange: [ 4, 7 ],
			repeatRoot: 4,
			stepIdx: 3
		});
	});

	it('looper changes steps predictably', () => {
		// Turn on looper
		comp.find('ForwardRef(Tooltip)').at(5).simulate('click', { shiftKey: true });
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				repeatRange: [ 4, 7 ],
				repeatRoot: 4,
				stepIdx: 3
			})
		});
	});

	it('modifies looper with subsequent shift clicks', () => {
		// Turn on looper
		comp.find('ForwardRef(Tooltip)').at(5).simulate('click', { shiftKey: true });
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				repeatRange: [ 4, 7 ],
				repeatRoot: 4,
				stepIdx: 3
			})
		});

		comp.find('ForwardRef(Tooltip)').at(10).simulate('click', { shiftKey: true });
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				repeatRange: [ 4, 12 ],
				repeatRoot: 4,
				stepIdx: 4 
			})
		});

		comp.find('ForwardRef(Tooltip)').at(0).simulate('click', { shiftKey: true });
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				repeatRange: [ 0, 12 ],
				repeatRoot: 0,
				stepIdx: 0 
			})
		});
	});

	it('clears looper when any node is clicked', () => {
		// Turn on looper
		comp.find('ForwardRef(Tooltip)').at(5).simulate('click', { shiftKey: true });
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, {
				repeatRange: [ 4, 7 ],
				repeatRoot: 4,
				stepIdx: 3
			})
		});

		comp.find('ForwardRef(Tooltip)').at(10).simulate('click');
		expect(setFoldState).toHaveBeenCalled();
		expect(setFoldState).toHaveBeenLastCalledWith({
			repeatRange: null,
			repeatRoot: -1,
			stepIdx: 12
		});
	});
});
