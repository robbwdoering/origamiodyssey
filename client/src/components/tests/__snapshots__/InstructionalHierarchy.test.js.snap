// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fold Editor renders without crashing 1`] = `ShallowWrapper {}`;
